aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+        else:
+            assert False, f"unsupported type: {t}"
+    elif isinstance(t, OptionalType):
+        # TODO: these arguments are smoothed over by the hacky wrapper
+        if argument and legacy_optional and str(t.elem) == 'Tensor':
+            if mutable:
+                return 'Tensor &'
+            else:
+                return 'const Tensor &'
+        if argument and str(t.elem) == 'Tensor':
+            if mutable:
+                return 'Tensor &'  # WUUUT
+            else:
+                return 'const c10::optional<Tensor>&'
+        elem = cpp_type(t.elem, mutable=mutable, argument=argument, legacy_optional=legacy_optional)
+        return f"c10::optional<{elem}>"
+    elif isinstance(t, ListType):
+        if str(t.elem) == 'bool':
+            assert t.size is not None
+            return f"std::array<bool,{t.size}>"
+        # TODO: remove this special case
+        if str(t.elem) == 'int' and argument:
+            return "IntArrayRef"
+        elif str(t.elem) == 'Tensor' and argument:
+            return "TensorList"
+        elif str(t.elem) == 'Tensor?' and argument and legacy_optional:
+            return "TensorList"
+        elif str(t.elem) == 'Dimname' and argument:
+            return "DimnameList"
+        elem = cpp_type(t.elem, mutable=mutable, argument=argument, legacy_optional=legacy_optional)
+        if argument:
+            # TODO: explicitly qualify namespace here
+            return f"ArrayRef<{elem}>"
+        else:
+            assert t.size is None, f"fixed size list returns not supported: {t}"
+            return f"std::vector<{elem}>"
+    else:
+        assert False
+
+def cpp_type_return(rs: Sequence[Argument]) -> str:
+    if len(rs) == 0:
+        return 'void'
+    elif len(rs) == 1:
+        return cpp_type(rs[0].type, mutable=rs[0].is_write, argument=False, legacy_optional=False)
+    else:
+        args = ','.join([cpp_type(r.type, mutable=r.is_write, argument=False, legacy_optional=False) for r in rs])
+        return f'std::tuple<{args}>'
+
+# Some simple code to exercise some of the functions we've been building
+def compute_type_derived_method_declarations(dispatch: str) -> List[str]:
+    type_derived_method_declarations: List[str] = []
+    for f in native_functions:
+        if f.dispatch is None or dispatch not in f.dispatch:
+            continue
+
+        name = str(f.func.name.name)
+        # TODO: delete this!
+        if f.func.is_out_fn():
+            name += '_out'
+        if f.func.name.overload_name:
+            name += f'_{f.func.name.overload_name}'
+
+        cpp_return = cpp_type_return(f.func.returns)
+
+        def format_arg(a: Argument) -> str:
+            return f"{cpp_type(a.type, mutable=a.is_write, argument=True, legacy_optional=True)} {a.name}"
+        cpp_args: List[str] = []
+        cpp_args.extend(map(format_arg, f.func.out_arguments))
+        cpp_args.extend(map(format_arg, f.func.arguments))
+
+        # Discover TensorOptions
```

Comment:
Should we be either more permissive or more proscriptive here? I.e. it might be more user friendly to either accept topt args in any order, or error if they appear out of order, instead of just letting out-of-order args escape the dragnet and cause unexpected behavior downstream. 

Aside: I think this is actually a good example of the kind of situation where a codegen pipeline needs a similar attitude to input handling as a compiler frontend - dumb/strict is fine, smart/lenient is fine, but dumb/lenient (i.e. fishing around for patterns and letting misses go by) is an easy way to introduce bad user experience, and IMO a major source of codegen's bad usability rep in general.

Reply:
The old codegen is much more proscriptive, requiring *specific* defaults and optionality before it matches. It is really wordy as a result. I could see this code being a bit too relaxed. I think fundamentally whether or not we grab arguments is a combination of type and name. However, the type is a relaxed test: I don't care if the type is optional or not in schema. If it is defaulted I need to generate defaulting code accordingly for tensor options.

Defo gonna change this.